lib/min_heap.rb

```diff
@@ -44,28 +53,52 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1) - Uncertain if Ruby under the hood implements an array in the strictest sense, then by inference it allocates a certain number of spaces to an array and retrieving the length would be constant. Otherwise it is O(n) due to check each element until it becomes nil.
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store.empty?
   end
 
   private
 
   # This helper method takes an index and
-  #  moves it up the heap, if it is less than it's parent node.
-  #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # moves it up the heap, if it is less than it's parent node.
+  # It could be **very** helpful for the add method.
+  # Time complexity: O(logn)
+  # Space complexity: Constant
   def heap_up(index)
-    
+    while index > 0
+      parent_index = ((index - 1) / 2)
+      if @store[parent_index].key > @store[index].key
+        swap(parent_index, index)
+        index = parent_index
+      else
+        return
+      end
+    end
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    child_index_left = ((index * 2) + 1)
+    child_index_right = ((index * 2) + 2)
+
+    #if there are no children, return
+    if @store[child_index_left].nil?
+      return 
+    elsif @store[child_index_right].nil?
+      swap(index, child_index_left)
```

Comment:
You should always compare the element at `index` with `child_index_left`.  This applies every place in the heap.